src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs

```diff
@@ -56,9 +60,24 @@ internal abstract class PEAssemblyBuilderBase : PEModuleBuilder, Cci.IAssemblyRe
 
         public override ISourceAssemblySymbolInternal SourceAssemblyOpt => _sourceAssembly;
 
-        internal override ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes()
+        internal override ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes(DiagnosticBag diagnostics)
         {
-            return _additionalTypes;
+            Interlocked.CompareExchange(ref _embeddedAttributesConstructors, CreateEmbeddedAttributesIfNeeded(diagnostics), null);
+
+            if (_embeddedAttributesConstructors.Count == 0)
+            {
+                return _additionalTypes;
+            }
+
+            var builder = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            builder.AddRange(_additionalTypes);
+
+            foreach (var constructor in _embeddedAttributesConstructors.Values)
```

Comment:
Good point. Switching to a deterministic data structure.

---
In reply to: [113571121](https://github.com/dotnet/roslyn/pull/18715#discussion_r113571121) [](ancestors = 113571121)